### ISSUE TITLE
Improve the default navigation flow for keyboard and screen reader users

### ISSUE THREAD

**Issue by danallsop** [ScottLogic/jumpstart-Leaflet#300]
> **Migrated from:** https://github.com/Leaflet/Leaflet/issues/7479
> **Original author:** danallsop
> **Original created:** 2021-02-24T14:27:53Z
---

## Motivation

The DOM order of elements affects the order in which screen readers announce content, and also the focus order.

I believe that the map controls should be available/presented to the user before other author provided content such as markers and popups.

This would reduce confusion and allow non-visual users to form a consistent mental model of the content. It also creates a more consistent and logical focus order even across different maps (perhaps by different authors) with varying content.

Relates to <span title="Web Content Accessibility Guidelines">WCAG</span> 2.1 Success Criterion [2.4.3 Focus Order](https://www.w3.org/TR/WCAG21/#focus-order) and [3.2.3 Consistent Navigation](https://www.w3.org/TR/WCAG21/#consistent-navigation).

## Proposed solution

<s>[Change `leaflet-control-container` to come before `leaflet-map-pane` in the DOM order.](https://github.com/Leaflet/Leaflet/issues/7479#changing-dom-order)</s>

<ins>[Arrow key navigation for focusable elements in `leaflet-map-pane`.](https://github.com/Leaflet/Leaflet/issues/7479#arrow-key-navigation)</ins>

## Additional context

danallsop raised a similar concern in https://github.com/Leaflet/Leaflet/issues/3210 (quoting the relevant parts):
> I was testing my website with a person who is completely blind and uses the VoiceOver (VO) function on an iPad. I was observing her as she browsed the entire webpage (through a tapping function in VO that mimics tabbing between elements*) and came across the Leaflet map.

> ## Expectations

> The next tab should land on the zoom controls, layer controls, and any other controls, so the user can understand what functions are available to control the map.

> The next tab should land on any marker that has a popup

**Comment by danallsop** [comment]
## Comments (3 total)

---
> **Comment by:** danallsop **on** 2021-11-21T23:22:27Z

> I believe that the map controls should be available/presented to the user before other author provided content such as markers and popups.

While I think this would be an improvement (especially when there are a lot of markers), the problem with this approach is that it causes disruptive tab stops between panning and interacting with the markers.

Below are descriptions of 3 different methods that I think should be considered. For each method, the scenario is a map with 100 markers.

<h2 id="current-situation"><a href="#current-situation">🔗</a> A. Current situation</h2>

1. <kbd>Tab</kbd> to map
2.  <kbd>Tab</kbd> through 100 markers
3. <kbd>Tab</kbd> to controls

In this situation, it would take more than 100 tab presses to reach the controls, to go into fullscreen for example. It's also difficult for users relying on screen readers to detect the available controls.

<h2 id="changing-dom-order"><a href="#changing-dom-order">🔗</a> B. Changing DOM order (<a href="https://github.com/Leaflet/Leaflet/pull/7796">[#7796](https://github.com/Leaflet/Leaflet/issues/7796)</a>)</h2>

If the DOM order of the map pane and controls is changed:

1. <kbd>Tab</kbd> to map
2. <kbd>Tab</kbd> through a few controls
3. <kbd>Tab</kbd> to markers

Now it's easy for keyboard users to use the controls but will have to <kbd>Tab</kbd> between map -> controls -> markers and <kbd>Shift</kbd> + <kbd>Tab</kbd> between markers -> controls -> map when they only want to pan and interact with markers. I.e. they're forced to tab past the controls each time.

<h2 id="arrow-key-navigation"><a href="#arrow-key-navigation">🔗</a> C. Arrow key navigation for focusable elements in <code>leaflet-map-pane</code></h2>

The optimal solution (that I can think of) would have to be something more like how e.g. Google's [marker accessibility](https://developers.google.com/maps/documentation/javascript/examples/marker-accessibility) works - markers are navigable using <kbd>arrow</kbd> keys:

1. <kbd>Tab</kbd> to map
2. <kbd>Tab</kbd> to markers (now use <kbd>arrow</kbd> keys if you want to navigate between markers)
3. <kbd>Tab</kbd> to controls<br>
  There are potentially 2 ways to achieve this:
    1. Implement something like the [ARIA Grid pattern](https://www.w3.org/TR/wai-aria-practices-1.1/#grid)
     (The more complex option)
    2. Set the `focusgroup` attribute on `leaflet-map-pane`, done!
     (Well, that's in theory, [`focusgroup` is a new proposal](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/Focusgroup/explainer.md) by Microsoft, which would need to be [polyfilled](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/Focusgroup/focusgroup_polyfill.js))

        Edit: new home for the `focusgroup` proposal: https://open-ui.org/components/focusgroup.explainer
---
> **Comment by:** danallsop **on** 2021-11-22T06:11:54Z

I think solution "C" would be good. But I don't think that the other maintainers allows the polyfill for `focusgroup`. Btw. with focusgroup it looks like [focusegroup='extend'](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/Focusgroup/explainer.md#632-extend---extending-the-focusgroup) would be needed too.

So we need to use the ARIA Grid pattern.
---
> **Comment by:** danallsop **on** 2021-12-09T13:07:58Z

> I think solution "C" would be good.

Just noting that arrow key navigation for markers would resolve a concern described in https://github.com/Leaflet/Leaflet/issues/3472#issue-76717597:

> as mourner said in [#1355](https://github.com/Leaflet/Leaflet/issues/1355) :
"an interesting accessibility problem... If the number of markers is too high, people will have big trouble tabbing through all the markers to get to other elements of a page"
